### PR TITLE
MB-8237 log user activation and deactivation

### PR DESCRIPTION
**Pull request by eamahanna**
## Description

This PR adds logging to the adminApi user, office user and admin user handlers. It also adds a function, `audit.CaptureAccountStatus`, and refactors a little bit of the audit.go file to make parts reusable. 

The output of the log for when a user is activated or deactivated is sent to cloudWatch and looks like: 
```
{
"level": "info",
    "ts": "2021-05-25T21:23:24.092Z",
    "caller": "audit/audit.go:107",
    "git_branch": "master",
    "git_commit": "1cfe3ab0a1d9728adcafe4893102fd6404ee8df1",
    "ecs_cluster": "arn:aws-us-gov:ecs:us-gov-west-1:015932076428:cluster/app-stg",
    "ecs_task_def_family": "app-stg",
    "ecs_task_def_revision": "1375",
    "milmove_trace_id": "8b15b762-67eb-4a6b-9b4d-43f0036e0553",
  "msg": "Audit Patch Office Users - account disabled"
  "event_type": "audit_patch_office_users", 
  "responsible_user_id": "4bbca4a9-2e97-4eab-8ee7-e29bc6c97b59", 
  "responsible_user_email": "20210602201721-8bf9bc264766@example.com", 
  "responsible_user_name": "Leo Spaceman", 
  "record_id": "144503a6-485c-463e-b943-d3c3bad11b09", 
  "record_type": "OfficeUser", 
  "record_created_at": "2021-05-27 18:32:56.571802 +0000 +0000", 
  "record_updated_at": "2021-06-04 17:56:01.82926 +0000 UTC m=+32.322890319", 
  "active_value": "false"
}
``` 

## Reviewer Notes

You can view this log in your local developer environment in your terminal.

I am working on adding another test for extractRecordInformation. I just wanted to get this PR in front of people :)

## Setup

* Start the admin app locally
* Navigate to an office user edit page
* Change the active value and hit save
* You should see the log in your terminal

## Code Review Verification Steps

* [ ] If the change is risky, it has been tested in experimental before merging.
* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] The requirements listed in [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely) have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-8237) for this change

### CODE REVIEW

File: pkg/services/audit/audit.go [hunk 1]
```diff
@@ -109,6 +71,89 @@ func Capture(model interface{}, payload interface{}, logger Logger, session *aut
...
+		zap.String("responsible_user_email", session.Email),
+	)
+
+	if session.IsAdminUser() || session.IsOfficeUser() {
+		logItems = append(logItems,
+			zap.String("responsible_user_name", fullName(session.FirstName, session.LastName)),
+		)
+	}
+	return logItems
+}
```
Comment (nelztruss) [inline pkg/services/audit/audit.go:118]: This could be more of a "pure function" if it did not accept the `logItems`... Really the responsibility would be to transform a `session` => `[]zap.Field`, leaving the (easy) responsibility of adding of the results to its destination aggregation.

File: pkg/services/audit/audit.go [hunk 2]
```diff
@@ -109,6 +71,89 @@ func Capture(model interface{}, payload interface{}, logger Logger, session *aut
...
+
+func extractRecordInformation(item reflect.Type, model interface{}, logItems []zap.Field) []zap.Field {
```
Comment (nelztruss) [inline pkg/services/audit/audit.go:120]: Same comment about making this a "pure function". This should also have the benefit of making it easier to test.

File: pkg/services/audit/audit.go [hunk 3]
```diff
@@ -109,6 +71,89 @@ func Capture(model interface{}, payload interface{}, logger Logger, session *aut
...
+	var logItems []zap.Field
+	eventType := extractEventType(request)
+	msg := flect.Titleize(eventType)
+	logItems = append(logItems, zap.String("event_type", eventType))
+
+	logItems = extractResponsibleUser(logItems, session)
+
+	item, err := validateInterface(model)
+	fmt.Print("TESTTEST")
+	fmt.Printf("%+v\n", model)
```
Comment (nelztruss) [inline pkg/services/audit/audit.go]: Test code left in.

**Review by nelztruss — COMMENTED** [review]
Just a couple of suggestions... Let me know if my explanations could be richer.

File: pkg/handlers/adminapi/admin_users.go [hunk 4]
```diff
@@ -171,6 +171,14 @@ func (h UpdateAdminUserHandler) Handle(params adminuserop.UpdateAdminUserParams)
 		return adminuserop.NewUpdateAdminUserInternalServerError()
 	}
 
+	// Log if the account was enabled or disabled (POAM requirement)
+	if payload.Active != nil {
+		_, err = audit.CaptureAccountStatus(updatedAdminUser, *payload.Active, logger, session, params.HTTPRequest)
+		if err != nil {
+			logger.Error("Error capturing account status audit record", zap.Error(err))
```
Comment (sandy-wright) [inline pkg/handlers/adminapi/admin_users.go:178]: Since we're doing the same thing in several different places, would you be willing to put the name of the handler in this log message?

File: pkg/services/audit/audit.go (hunk 1 above)
Comment (sandy-wright) [reply pkg/services/audit/audit.go:118]: So this would mean using this function would look like:
```go
logItems = append(logItems, extractResponsibleUser(session))
```
instead? Just to make sure I understand what you mean 😄

File: pkg/services/audit/audit_test.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ func TestCapture(t *testing.T) {
 		adminUserID, _ := uuid.FromString(uuidString)
 
 		session := auth.Session{
-			AdminUserID: adminUserID,
+			UserID: adminUserID,
```
Comment (sandy-wright) [inline pkg/services/audit/audit_test.go:37]: Why did you need to change this? 🤔

File: pkg/services/audit/audit_test.go [hunk 6]
```diff
@@ -164,3 +164,132 @@ func TestCapture(t *testing.T) {
...
+		CreatedAt: time.Now(),
+		UpdatedAt: time.Now(),
+	}
+	logger := zap.NewNop()
+
+	uuidStringAdmin := "4ad12fe7-1514-4b6b-a35d-ce68e6c5b1fc"
+	adminUserID, _ := uuid.FromString(uuidStringAdmin)
+
+	session := auth.Session{
+		AdminUserID: adminUserID,
```
Comment (sandy-wright) [inline pkg/services/audit/audit_test.go:182]: Why is it now `AdminUserID` here but `UserID` up above?

File: pkg/services/audit/audit_test.go [hunk 7]
```diff
@@ -164,3 +164,132 @@ func TestCapture(t *testing.T) {
...
+			}
+			if field.Key == "active_value" {
+				activeValue = field.String
+			}
+		}
+
+		if assert.NotEmpty(t, zapFields) {
+			assert.Equal(t, "event_type", zapFields[0].Key)
+			assert.Equal(t, "audit_post_admin_users", eventType)
+			assert.Equal(t, "active_value", zapFields[8].Key)
```
Comment (sandy-wright) [inline pkg/services/audit/audit_test.go:231]: Wow, I wish this were a map instead because it makes me so uncomfortable to see a random `8` as an index 😅. I don't think you need to change that though

File: pkg/services/audit/audit.go (hunk 1 above)
Comment (nelztruss) [reply pkg/services/audit/audit.go:118]: Almost, gotta add the `...` operator for variadic expansion:
```
logItems = append(logItems, extractResponsibleUser(session)...)
```

https://play.golang.org/p/xRUfmAtYSAX

File: pkg/services/audit/audit_test.go (hunk 7 above)
Comment (nelztruss) [reply pkg/services/audit/audit_test.go:231]: I agree with Sandy... You really just want to test membership, e.g. if it "hasA"; you don't really care about what order it comes in. If someone else eventually adds another field in there somewhere, this test now breaks, meaning this is a bit of a brittle construct.

File: pkg/services/audit/audit.go (hunk 1 above)
Comment (sandy-wright) [reply pkg/services/audit/audit.go:118]: oooOOoh, I haven't had to use that syntax yet!

Comment (eamahanna) [reply pkg/services/audit/audit.go:118]: That makes sense! fixed

File: pkg/services/audit/audit.go (hunk 2 above)
Comment (eamahanna) [reply pkg/services/audit/audit.go:120]: fixed!

File: pkg/services/audit/audit_test.go (hunk 5 above)
Comment (eamahanna) [reply pkg/services/audit/audit_test.go:37]: It was a typo. I was messing around with the test and forgot to change it back. Reverted!

File: pkg/services/audit/audit_test.go (hunk 6 above)
Comment (eamahanna) [reply pkg/services/audit/audit_test.go:182]: See other comment response above :)

File: pkg/services/audit/audit_test.go (hunk 7 above)
Comment (eamahanna) [reply pkg/services/audit/audit_test.go:231]: Sounds good. I will see if I can make this less brittle.

File: pkg/services/audit/audit.go [hunk 8]
```diff
@@ -109,6 +71,88 @@ func Capture(model interface{}, payload interface{}, logger Logger, session *aut
 	return logItems, nil
 }
 
+// CaptureAccountStatus captures an audit record when a user account is enabled or disabled
+func CaptureAccountStatus(model interface{}, activeValue bool, logger Logger, session *auth.Session, request *http.Request) ([]zap.Field, error) {
+	var logItems []zap.Field
+	eventType := extractEventType(request)
```
Comment (sandy-wright) [inline pkg/services/audit/audit.go:77]: Can we change this value to something distinct? If we change an office user without changing the active status, we'll have the exact same event type. I think they should be different so that we can filter the logs for these events. Maybe we can have one eventType for all user types - like `audit_change_active_status`? Alternatively, we can just tack something onto the end of the generated type - `extractEventType(request) + "_active_status"`

File: pkg/services/audit/audit_test.go [hunk 9]
```diff
@@ -78,7 +78,7 @@ func TestCapture(t *testing.T) {
 		}
 
 		session := auth.Session{
-			AdminUserID: adminUserID,
+			UserID: adminUserID,
```
Comment (sandy-wright) [inline pkg/services/audit/audit_test.go:81]: Another one of those typos :)

**Review by sandy-wright — CHANGES_REQUESTED** [review]
This looks really good to me, but I think we do need to differentiate the event types here. We want to be able to filter for these events in particular, so a specific event type would be super helpful.

Other than that, it worked really smoothly in testing. Let me know when you fix that bit and I'm good to approve!

File: pkg/services/audit/audit.go (hunk 8 above)
Comment (eamahanna) [reply pkg/services/audit/audit.go:77]: Done!

File: pkg/services/audit/audit_test.go [hunk 10]
```diff
@@ -164,3 +164,119 @@ func TestCapture(t *testing.T) {
...
+
+	t.Run("Sucessfully logs account disabled", func(t *testing.T) {
+		zapFields, _ := CaptureAccountStatus(&model, false, logger, &session, req)
+
+		fieldsMap := map[string]string{}
+		for _, f := range zapFields {
+			fieldsMap[f.Key] = f.String
+		}
+
+		if assert.NotEmpty(t, zapFields) {
```
Comment (sandy-wright) [inline pkg/services/audit/audit_test.go:214]: Should this check be done before you create the map?

**Review by sandy-wright — APPROVED** [review]
LGTM! Thanks for making those changes! One last comment on the test, but it's a small thing
